Update to use TheTVDB API v2

palortoff opened the pull request.

palortoff commented:
@edwellbrook 
I'm pretty much done with the 2.0 implementation. 
Please have look and let me know if I need to change anything.

Do you prefer the commits all squashed into one?

edwellbrook commented:
Hey! Thanks so much for your work on this! So excited to get this merged in.

I'm currently without a computer so please bear with me to get this looked at and for me to leave my feedback. Hope it won't take longer than a couple days :)

Thanks again, so much, for your work on this. Will review ASAP!

edwellbrook commented:
(didnt see the additional message at the bottom, sorry)

I'm fine with many commits like this, but I always have the ability to squash when I merge if needed :)

edwellbrook commented on wercker.yml:
```diff
@@ -1,4 +1,4 @@
-box: node:4.0.0
+box: node:6.0.0
```
Why's this needed? Happy to bump if code really needs it, but if not can we move it to a separate PR to keep this one super focused.

palortoff replied:
added separate PR.
needed for default parameters.

edwellbrook commented on package.json:
```diff
@@ -20,15 +20,21 @@
-    "request": "2.74.0",
-    "xml2js": "0.4.17"
+    "debug": "^2.2.0",
+    "lodash": "^4.15.0",
```
Conscious of adding many dependencies here. Don't mind lodash being added as a dev dep for tests, however, if we're only using it for two functions in the actual library can we either do things ourselves, or require the exact packages (`require('lodash/defaults')`).

for reference: http://knpw.rs/blog/using-lodash

palortoff replied:
Removed lodash usage where possible (tests).
index.js uses `defaults` and `flatten`. Changed these into `require('lodash/defaults')`. I prefer not to reinvent the wheel here and stick to using lodash, but including as little as needed.

edwellbrook replied:
Sounds good to me!

edwellbrook commented on .eslintignore:
```diff
@@ -0,0 +1,2 @@
+playground
+legacy
```
Could we move `.eslint` files to a separate PR. They're great, but I want to keep this PR as focused as possible.

Additionally can we have trailing new lines, please.

palortoff replied:
made separate PR.
set `no-multiple-empty-lines / maxEOF` to 10

edwellbrook commented on package.json:
```diff
@@ -20,15 +20,21 @@
-    "jszip": "3.1.2",
-    "request": "2.74.0",
-    "xml2js": "0.4.17"
+    "debug": "^2.2.0",
```
Not convinced the information being debugged is useful enough to warrant outputting, and an additional dependency. Can we remove this please.

palortoff replied:
It was useful during development. 
I chose debug over console log, as there is no real harm in leaving it in, but I'll remove it now.

edwellbrook commented on package.json:
```diff
@@ -20,15 +20,21 @@
+    "debug": "^2.2.0",
+    "lodash": "^4.15.0",
+    "request": "^2.74.0",
+    "request-promise": "^4.1.1"
```
Could we use `request-promise-core` instead? Should be the same right? And means we'll use native promises and `bluebird` as a sub dep.

palortoff replied:
`request-promise-core` is not recommended as a dependency other that for a `request-promise-x` implementation.

Please see http://programmers.stackexchange.com/questions/278778/why-are-native-es6-promises-slower-and-more-memory-intensive-than-bluebird .

I suggest not to sacrifice better performance and memory usage to minimizing dependencies.

edwellbrook replied:
In this case I'm happy to agree with you. Thanks for looking into this and explaining the situation to me!

palortoff commented:
Hi,

no worries. Good things come to those who wait ;-)

I think I made all the changes you requested (except for the `request-promise-core` one, see comment).

If you merge the other two PRs (and then update the `new-tvdb-api` branch) I'll rebase this PR. If you don't object I would also squash all the commits into one then.

edwellbrook commented:
Hey made some changes and tried to tidy up the branches. `new-tvdb-api` has been reset to `master` + your two other pull requests. If you could please rebase your pull request with `new-tvdb-api` that'd be great.

Some additonal feedback on this PR: Could we add the comments above methods back in. The library previously described the methods and parameters with types, but these have now been removed with your changes. Would be good to have those back in.

palortoff commented:
All previous commits have been squashed into one. All rebased onto your `new-tvdb-api`.

Added comments for type info. Usually I am too lazy to do that, but with an API library, that might be helpful ;-)

edwellbrook commented on README.md:
```diff
@@ -2,21 +2,17 @@
-- Use zip data instead of straight xml where possible
+- Set language at initialisation or on function call
+- Return values through promises (dropped callback support)
+- Use new JSON api from TheTVDB
```
Capitalisation (api -> API)

edwellbrook commented on README.md:
```diff
@@ -2,21 +2,17 @@
-- Supports both node callback functions and promises
-- Utility function to parse TheTVDB API's pipe list (e.g. "|Name|Name|Name|Name|")
-- Use zip data instead of straight xml where possible
+- Set language at initialisation or on function call
```
on function call -> on each function call

edwellbrook commented on index.js:
```diff
@@ -7,440 +8,217 @@
+    /**
+     * @param {String} url
+     */
+    static setBaseUrl(url) {
```
Not convinced this is needed right now. Can we remove this method and initialise the base URL on line 15 with a `const` thanks.

edwellbrook replied:
Feel free to open as another pull request if you really think this should be included. Thanks!

edwellbrook commented on index.js:
```diff
@@ -7,440 +8,217 @@
-};
-
-module.exports = Client;
+module.exports = Client;
```
New line here.

edwellbrook commented on index.js:
```diff
@@ -7,440 +8,217 @@
-
-function parsePipeList(list) {
-    return list.replace(/(^\|)|(\|$)/g, "").split("|");
+function hasNextPage(response) {
```
This is missing a function comment. I believe this is a private helper method, correct?

edwellbrook commented on index.js:
```diff
@@ -7,440 +8,217 @@
+    /**
+     * @private
+     */
+    requestGet(options, language) {
```
Private methods still need descriptions and parameters listed in the jsdoc comment

edwellbrook commented on index.js:
```diff
@@ -7,440 +8,217 @@
+
+                    json: true,
+                    headers: {
+                        "User-Agent": "edwellbrook/node-tvdb",
```
This is an invalid user agent. Should be `node-tvdb/{version}`.

edwellbrook replied:
Also odd that we're now using single quotes for strings everywhere but here?

palortoff replied:
I copied it from the original new-tvdb-api branch...

palortoff commented:
I added the URLs to the API documentation in the function documentation on the README.md.

Adding the description and/or thetvdb API doc URL to the index.js has not much value in my opinion as the package is already well documented in the README.md.
Adding the description and URLs to the function comment would violate DRY. If you think more documentation in the function comment is needed I suggest generating either the README.md or the function comments using a gulp script prior to release.

Also I intentionally only added type information to the public functions and marked private functions as private. This documentation has value to the user as their IDE will use it to provide type hints and prevent usage of private functions. There is no need to provide type information for private functions. I also did not add a comment to the private helper function as it is not exported.

I strongly believe that over-documenting does more harm than good. Documentation needs to be maintained and quickly deviates from the code itself if not maintained well. There is a need of some documentation, especially when providing a public API, but I think we've met that need in the README.md.

If you disagree with me you are free to add the documentation wherever you deem it necessary.

edwellbrook commented:
Hey, very much thinking of other people wanting to comtribute to the library. In this use case, its incredibly important to have a clearly documented code. The README is fantastic for end users, but as a developer of the library, I need type information for private methods, and really want descriptions and URLs inline with the code.

I agree there is a level of duplication with this, but its largely unavoidable at this point as we're trying to cater for both end users as well as developers of this library.

Sorry it's taking a while to work through this but I'll be without a computer for a few more days. As this PR is against a dev branch I'll be more accepting, however it will still need some tidying (points raised previously) before merging into master. Thanks as always for your patience in getting this accepted and merged.

palortoff commented:
Hi,

I added documentation to private functions.
For detailed documentation on @public functions containing description, urls and example code please look at #63.
Here I moved the function documentation from the README.md to index.js and added a script to generate the parts in the README.md.

This way we meet your need for documentation with the code and also avoid duplication.

I hope this solution is OK with you :-)
